### DISCUSSION TITLE
Trigger Socket IO Emit with PHP

### DISCUSSION THREAD

**Discussion by El-Tommy** [socketio/socket.io#4404] [Q&A]
I have an old project with Socket.IO 1 that I must migrate to Socket.IO 4.

It happens that I use the [Elephant.IO](https://wisembly.github.io/elephant.io/#usage) library in PHP to call the events available on the server with Socket IO.

It so happens that the Elephant.io project has been discontinued and Socket.IO support ended in version 2.

Another [project](https://github.com/rase-/socket.io-php-emitter) recommended by the Socket.IO website references a library that hasn't had updates for 8 years and stopped support in Socket.IO 3.

What should I currently use to fire events to the Socket.IO server from PHP Server?

I've read in a few places that one option would be to have RabbitMQ, but I don't think it's good to have another intermediary server if I can communicate directly with SocketIO.

**Comment by qiulang** [comment]
I would like to share my experience as I have done the same thing myself.

First, [Ratchet](https://github.com/ratchetphp/Ratchet) is a popular php WebSocket Server, checking its document you can see it uses  [ZeroMQ](https://zeromq.org/) to make ws server talk to http server.

I don't want to use ZeroMQ so I try some other pub sub tools.  RabbitMQ is definitely an option but I don't want to use it either. So finally I resort to redis pub sub. For me it has lowest learning curve and dependency. So far it works well.

If you want  to "communicate **directly** with SocketIO" you may consider zeromq but I don't think an "intermediary server" is necessarily a bad thing. For us we already use redis, so it fits the bill.

**Reply by El-Tommy** [reply]
Hi @qiulang,

As I understand Ratchet would replace SocketIO as websocket server, right?
If so, it wouldn't be my case, as I already use SocketIO with other applications (like Angular or JavaScript).

You mentioned Redis. As I understand it, correct me if I'm wrong, you "emit" the event to Redis and it is in charge of issuing it to Socket.IO, is that it?

**Reply by qiulang** [reply]
The reason I mentioned Rachet is to let you know how ws server can communicate web server. I also use socket.io to develop my ws server. As there are 2 independent servers here, ws server & web server, using pub-sub to let them communicate is a nature selection (at least in my case). You may further check redis pub-sub, e.g. https://redis.com/redis-best-practices/communication-patterns/pub-sub/

**Comment by tohenk** [comment]
Try maintained [Elephant.IO](https://github.com/ElephantIO/elephant.io).
